docs/glossary.md

```diff
@@ -0,0 +1,71 @@
+
+It is recommended to avoid using popular programming language terms such as "field" or "property" to describe an element of an object in OpenTofu.
+
+Reference: [link](./diagnostics.md#diagnostic-description-writing-style)
+
+## Data sources/data resource
+* "Data resource" refers to a block of type data, and the associated object it declares.
+* "Data source" is the remote thing that the data resource reads from.
+* "Data resource type" is what is represented by the first label in a data block header, and the associated declarations and code for it in the provider plugin.
+
+Reference: [link](https://github.com/opentofu/opentofu/pull/3389#discussion_r2440264786)
+
+## Diagnostic
+"Diagnostics" is the general term we use to describe the error and warning
+messages that OpenTofu returns when there are problems with the configuration,
+or when interactions with external systems fail.
+
+Reference: [link](./diagnostics.md)
+
+## Expression
+Expressions refer to or compute values within a configuration. The simplest expressions are just literal values, like "hello" or 5, but the OpenTofu language also allows more complex expressions such as references to data exported by resources, arithmetic, conditional evaluation, and a number of built-in functions.
+
+Reference: [link](https://opentofu.org/docs/language/expressions/)
+
+## Evaluation context
+A set of already known functions, input values, local values, resources, etc. that is used to evaluate an expression that can reference any of the concepts listed above.
+
+The list of concepts above, in the context of HCL evaluation, are called [variables](#variable-hcl).
+
+## Mark/value mark
+This is an implementation detail term used to describe an attribute that a particular value can have.
+Instead of directly using the term "mark", it is encouraged to use terms like "sensitive value" or "ephemeral value".
+
+Reference: [link](./diagnostics.md#diagnostic-description-writing-style)
+
+## Resource/resource instance/resource type
+* Resource - is what is declared by a `resource`, `data` or `ephemeral` block.
+* Resource instance - is what such a block can declare zero or more of, when using the `count`, `for_each`, or `enabled` arguments.
+* Resource type - the type of a "resource". E.g., `aws_instance` is a "resource type".
+
+It is recommended to use the following terms when discussing about "resource" blocks:
+* managed resource - a block declared as `resource "type" "name" {}`
+* data resource - a block declared as `data "type" "name" {}`
+* ephemeral resource - a block declared as `ephemeral "type" "name" {}`
+
+Reference: [link](./diagnostics.md#diagnostic-description-writing-style)
+
+## Unknown value/Computed value
+Values that are generated by evaluating an expression that can most of the time only be done during or after the apply phase.
+
+Reference: [link](./diagnostics.md#diagnostic-description-writing-style)
+
+## Variable (HCL)
```

Comment:
Would recommend moving this closer to EvalContext (HCL)

Reply:
Glossary, AFAIK, is alphabetically ordered. If we want to move this variable closer to the eval context then we should create groups then? WDYT?